### ISSUE TITLE
SVI.run on GPU affects subsequent executions on CPU of with a new instance of SVI

### ISSUE THREAD

**Issue by muupan** [pyro-ppl/numpyro#1496]
I asked about this issue in google/jax's discussions (https://github.com/google/jax/discussions/12945) but was not able to get an answer, so now I am asking here.

The issue is, when I run `SVI.run` of some SVI object on CPU by setting the default device of jax to CPU, the result is affected by whether I ran `SVI.run` of some other SVI object on GPU before. I expect the results of executions of the same code must be same as long as they are on the same device, but there is somehow a difference between pre-GPU and post-GPU results. I suspect the issue is related to JIT as the difference disappears when I turn off JIT by setting JAX_DISABLE_JIT=1.

Is this a bug or am I missing something? Are there anything I can do to make sure CPU results are not affected by GPU executions with JIT enabled?

Here is how to reproduce the issue.
```python
import jax
import jax.numpy as jnp
import numpy as np
import numpyro
import numpyro.distributions as dist
import numpyro.contrib.control_flow


def model(data):
    t_train, n = data.shape

    mu = numpyro.param(
        "mu",
        init_value=jnp.ones(n),
    )

    def transition(carry, t):
        h_prev, u_prev = carry
        h_t = 0.1 * (u_prev ** 2) + h_prev
        scale = jnp.sqrt(h_t)
        y_t = numpyro.sample(
            "y",
            dist.Normal(loc=mu, scale=scale),
        )
        u_t = y_t - mu
        return (h_t, u_t), y_t

    h_0 = jnp.ones(n) * 0.01
    u_0 = jnp.ones(n)

    carry_init = (h_0, u_0)

    with numpyro.handlers.condition(data={"y": data[1:]}):
        numpyro.contrib.control_flow.scan(
            transition, carry_init, jnp.arange(1, t_train)
        )


def f():
    numpyro.util.set_rng_seed(0)

    data = np.random.normal(loc=0, scale=0.01, size=(20, 1))
    data_jnp = jnp.array(data)

    def guide(*args, **kwargs):
        return None

    optimizer = numpyro.optim.Adam(step_size=1.0)
    svi = numpyro.infer.SVI(model, guide, optimizer, loss=numpyro.infer.Trace_ELBO())
    svi_result = svi.run(
        jax.random.PRNGKey(0),
        1000,
        data_jnp,
        progress_bar=True,
    )
     
    return [float(x) for x in svi_result.params.values()]


with jax.default_device(jax.devices("cpu")[0]):
    y1 = f()

with jax.default_device(jax.devices("cpu")[0]):
    y2 = f()

with jax.default_device(jax.devices("gpu")[0]):
    y3 = f()

with jax.default_device(jax.devices("cpu")[0]):
    y4 = f()

print(f"{y1=}")
print(f"{y2=}")
print(f"{y3=}")
print(f"{y4=}")
```
Output:
```
100%|██████████| 1000/1000 [00:02<00:00, 422.04it/s, init loss: 32.2299, avg. loss [951-1000]: -3.3998]
100%|██████████| 1000/1000 [00:02<00:00, 443.21it/s, init loss: 32.2299, avg. loss [951-1000]: -3.3998]
100%|██████████| 1000/1000 [00:07<00:00, 126.88it/s, init loss: 32.2299, avg. loss [951-1000]: -3.4205]
100%|██████████| 1000/1000 [00:01<00:00, 894.96it/s, init loss: 32.2299, avg. loss [951-1000]: -3.4193]
y1=[0.05775194615125656]
y2=[0.05775194615125656]
y3=[0.04599900171160698]
y4=[-0.03762827441096306]
```
I expect `y1=y2=y4` as they are all executed on CPU, but `y4`, computed after the GPU execution of `y3`, is different from `y1` and `y2`, which is the issue.

I'm using jax==0.3.21, jaxlib==0.3.20+cuda11.cudnn8, and numpyro==0.10.1.

**Comment by muupan** [comment]
Can anyone reproduce the result? I have trouble in running both cpu and gpu unit tests using numpyro as the execution order matters because of the issue. I have no clue what is causing it.

**Comment by fehiepsi** [comment]
Hi @muupan, this seems to be an issue but I'm not sure how to address it. Does this only happen with numpyro? Does this only happen when using `scan`? It would be nice to make a small reproducible code to isolate the issue, e.g. to create a stochastic loss function and optimize the parameters using optax. This will be helpful for us and jax team to see what's going on.

**Comment by muupan** [comment]
Thanks for your response!

> Does this only happen with numpyro?

I have experienced this issue only with numpyro. I am new to jax and numpyro so not sure if this issue is specific to numpyro.

> Does this only happen when using scan?

I have not experienced this issue without scan. The code I shared is minimal code to reproduce the issue at the moment. If I can make it smaller, I will post it too.

Additionally, I tried disabling JIT for scan. There remains a difference between pre-GPU and post-GPU results.
```
@@ -31,9 +31,10 @@ def model(data):
     carry_init = (h_0, u_0)

     with numpyro.handlers.condition(data={"y": data[1:]}):
-        numpyro.contrib.control_flow.scan(
-            transition, carry_init, jnp.arange(1, t_train)
-        )
+        with jax.disable_jit():
+            numpyro.contrib.control_flow.scan(
+                transition, carry_init, jnp.arange(1, t_train)
+            )
```
```
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 1095.81it/s, init loss: 32.2299, avg. loss [951-1000]: -3.3908]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 1099.44it/s, init loss: 32.2299, avg. loss [951-1000]: -3.3908]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:01<00:00, 664.08it/s, init loss: 32.2299, avg. loss [951-1000]: -3.4292]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 1202.78it/s, init loss: 32.2299, avg. loss [951-1000]: -3.3934]
y1=[-0.04256464168429375]
y2=[-0.04256464168429375]
y3=[-0.004527210257947445]
y4=[-0.0427229106426239]
```

**Comment by fehiepsi** [comment]
I convert the model into jax code. You can use this to report to jax team
```python
import optax

def loss_fn(params, data):
    t_train, n = data.shape
    mu = params["mu"]

    def transition(carry, t):
        h_prev, u_prev = carry
        h_t = 0.1 * (u_prev ** 2) + h_prev
        scale = jnp.sqrt(h_t)
        y_t = data[t]
        u_t = y_t - mu
        return (h_t, u_t), (mu, scale, y_t)

    h_0 = jnp.ones(n) * 0.01
    u_0 = jnp.ones(n)
    carry_init = (h_0, u_0)
    _, (mu, scale, y_t) = jax.lax.scan(
        transition, carry_init, jnp.arange(1, t_train))
    normalize_term = jnp.log(jnp.sqrt(2 * jnp.pi) * scale)
    value_scaled = (y_t - mu) / scale
    return (0.5 * value_scaled**2 + normalize_term).sum()

def f():
  np.random.seed(0)
  optimizer = optax.adam(1.0)
  data = np.random.normal(loc=0, scale=0.01, size=(20, 1))
  data_jnp = jnp.array(data)
  params = {"mu": jnp.ones(data_jnp.shape[-1])}
  opt_state = optimizer.init(params)

  @jax.jit
  def step(params, opt_state):
    loss_value, grads = jax.value_and_grad(loss_fn)(params, data_jnp)
    updates, opt_state = optimizer.update(grads, opt_state, params)
    params = optax.apply_updates(params, updates)
    return params, opt_state, loss_value

  for i in range(1000):
    params, opt_state, loss_value = step(params, opt_state)
    if i % 100 == 0:
      print(f'step {i}, loss: {loss_value}')

  return params
```

**Comment by muupan** [comment]
Thanks for your code! I removed optax so the code depends only on jax and numpy.
```python
import jax
import jax.example_libraries.optimizers
import jax.numpy as jnp
import numpy as np


def loss_fn(params, data):
    t_train, n = data.shape
    mu = params["mu"]

    def transition(carry, t):
        h_prev, u_prev = carry
        h_t = 0.1 * (u_prev ** 2) + h_prev
        scale = jnp.sqrt(h_t)
        y_t = data[t]
        u_t = y_t - mu
        return (h_t, u_t), (mu, scale, y_t)

    h_0 = jnp.ones(n) * 0.01
    u_0 = jnp.ones(n)
    carry_init = (h_0, u_0)
    _, (mu, scale, y_t) = jax.lax.scan(transition, carry_init, jnp.arange(1, t_train))
    normalize_term = jnp.log(jnp.sqrt(2 * jnp.pi) * scale)
    value_scaled = (y_t - mu) / scale
    return (0.5 * value_scaled ** 2 + normalize_term).sum()


def f():
    np.random.seed(0)
    optimizer = jax.example_libraries.optimizers.adam(1.0)
    data = np.random.normal(loc=0, scale=0.01, size=(20, 1))
    data_jnp = jnp.array(data)
    params = {"mu": jnp.ones(data_jnp.shape[-1])}
    opt_state = optimizer[0](params)

    @jax.jit
    def step(i, opt_state):
        params = optimizer[2](opt_state)
        loss_value, grads = jax.value_and_grad(loss_fn)(params, data_jnp)
        opt_state = optimizer[1](i, grads, opt_state)
        return opt_state, loss_value

    for i in range(1000):
        opt_state, loss_value = step(i, opt_state)
        if i % 100 == 0:
            print(f"step {i}, loss: {loss_value}")

    params = optimizer[2](opt_state)

    return params


with jax.default_device(jax.devices("cpu")[0]):
    y1 = f()

with jax.default_device(jax.devices("cpu")[0]):
    y2 = f()

with jax.default_device(jax.devices("gpu")[0]):
    y3 = f()

with jax.default_device(jax.devices("cpu")[0]):
    y4 = f()

print(f"{y1=}")
print(f"{y2=}")
print(f"{y3=}")
print(f"{y4=}")
```
```
step 0, loss: 32.22989273071289
step 100, loss: -3.4972240924835205
step 200, loss: -3.4972896575927734
step 300, loss: -3.497291088104248
step 400, loss: -3.097050905227661
step 500, loss: -3.3995137214660645
step 600, loss: -3.49708890914917
step 700, loss: -3.4879231452941895
step 800, loss: -3.485825538635254
step 900, loss: -3.4970479011535645
step 0, loss: 32.22989273071289
step 100, loss: -3.4972240924835205
step 200, loss: -3.4972896575927734
step 300, loss: -3.497291088104248
step 400, loss: -3.097050905227661
step 500, loss: -3.3995137214660645
step 600, loss: -3.49708890914917
step 700, loss: -3.4879231452941895
step 800, loss: -3.485825538635254
step 900, loss: -3.4970479011535645
step 0, loss: 32.22989273071289
step 100, loss: -3.4972550868988037
step 200, loss: -3.4972901344299316
step 300, loss: -3.497291088104248
step 400, loss: -3.4216434955596924
step 500, loss: -3.4793431758880615
step 600, loss: -3.4967548847198486
step 700, loss: -3.4921681880950928
step 800, loss: -3.463667154312134
step 900, loss: -3.49725079536438
step 0, loss: 32.22989273071289
step 100, loss: -3.4972240924835205
step 200, loss: -3.4972896575927734
step 300, loss: -3.497291088104248
step 400, loss: -3.415701389312744
step 500, loss: -3.485470771789551
step 600, loss: -3.4956696033477783
step 700, loss: -3.494507074356079
step 800, loss: -3.457836627960205
step 900, loss: -3.4971232414245605
y1={'mu': DeviceArray([0.05775195], dtype=float32)}
y2={'mu': DeviceArray([0.05775195], dtype=float32)}
y3={'mu': DeviceArray([0.045999], dtype=float32)}
y4={'mu': DeviceArray([-0.03762827], dtype=float32)}
```

**Comment by fehiepsi** [comment]
Nice! I guess we can close this now because the issue is not related to numpyro. :)
